File: services/api/app/domain/pair_notebook.py
"""The two-person notebook: consent, its cycle, and what closing it costs.

Spec «Nếp truyền giấy» sections 6 and 7, ADR-0027 K1, K2, K6.

## Why consent is a proposal plus grants, and not a level

The first design had `consent_level: int`, and a number is exactly the shape
that lets a later feature quietly qualify: 2 means «notebook», 3 means «couple»,
and the code that reads `>= 2` has already granted the thing 3 means. Here a
consent is (purpose, person, terms version, cycle), each purpose is asked for
separately, and a lower rung never implies a higher one. Reading a grant is
reading a row that names what was agreed to.

## Why a cycle

Closing the notebook does not delete it (section 7.6: «đóng sổ là đóng» --
sheets that were open are cancelled, plans that stood stay readable). Opening
another one later is a new agreement, not a revival of the old, so consent is
scoped to a cycle and a closed cycle's grants never apply to the next one.

Pure functions over dicts. No I/O, no ORM, no framework.
"""

from __future__ import annotations

import hashlib
from datetime import datetime, timedelta

from app.domain import interests, pair_paper

__all__ = [
    "CONSENT_PURPOSES",
    "CONSTRAINT_KINDS",
    "CYCLE_STATES",
    "NotebookError",
    "PER_PERSON_PURPOSES",
    "can_bat_doi",
    "chat_consent_active",
    "dang_cho",
    "granted_by",
    "granted_purposes",
    "gu_hai_nguoi",
    "nguoi_lo_suy",
    "vai_tuan",
    "han_de_nghi",
    "xem_truoc_dong_so",
]

#: How long an unanswered offer stands. A week, so that «I will think about it»
#: survives one of them being away, and so that an offer nobody answered stops
#: being an offer rather than sitting in the notebook forever waiting to be
#: accepted by somebody who has forgotten it was asked.
HAN_DE_NGHI = timedelta(days=7)

#: `pending` until both have granted `lap_so`; `closed` is final for that
#: cycle. The CHECK on `pair_notebook_cycles.state` spells this again.
CYCLE_STATES = ("pending", "active", "closed")

#: The consent ladder of section 6.1, tier 2 upward. Tier 1 («nhận lời đi
#: chơi») is the invitation itself and has no row. `doc_chat` is tier 4 and is
#: off until somebody turns it on: there is no default that reads a chat.
#: `chia_gu` (ADR-0034) is not a rung both climb: each person turns it on for
#: THEMSELVES -- «let the other see my taste, and let Nếp use it here» -- and a
#: proposal for it has one answer, its proposer's. It sits in this tuple so the
#: per-person switches (`my_consents`, `their_consents_granted`) carry it.
CONSENT_PURPOSES = ("lap_so", "bat_doi", "doc_chat", "chia_gu")

#: Purposes one person decides alone (ADR-0034 §2.1).
PER_PERSON_PURPOSES = ("chia_gu",)

#: The two shared constraints of section 6.4. Two, not a free list: a list
#: grows into a profile, and this is meant to stay the smallest thing that
#: keeps a draft from being wrong about somebody.
CONSTRAINT_KINDS = ("khong_an_duoc", "dung")


class NotebookError(Exception):
    def __init__(self, code: str):
        super().__init__(code)
        self.code = code


def han_de_nghi(now: datetime) -> datetime:
    """When an offer made at `now` lapses."""
    return now + HAN_DE_NGHI


def _live(consent: dict, *, now: datetime | None) -> bool:
    """A grant that is still in force: granted, not revoked, not expired.

    The expiry is the PROPOSAL's, not the grant's. An offer nobody answered
    within its window stops being an offer; a grant that was answered stands
    until it is revoked. «Answered» is the proposal's `completed_at`: before
    2026-09-23 the window was applied to completed proposals too, so an agreed
    «Một đôi» would have switched itself off a week later.
    """
    if not consent.get("granted_at"):
        return False
    if consent.get("revoked_at"):
        return False
    if consent.get("proposal_completed_at") is not None:
        return True
    expires_at = consent.get("proposal_expires_at")
    if expires_at is not None and now is not None and now >= expires_at:
        return False
    return True


def granted_by(
    consents: tuple[dict, ...] | list[dict],
    person_id: str,
    *,
    now: datetime | None = None,
) -> frozenset[str]:
    """What ONE person has granted and not taken back.

    Never the answer to «may this happen» -- that is `granted_purposes`, and
    both people are part of it. This is only for showing a person their own
    switches, and for showing each of them which of the two has not answered
    yet: a screen that said «đang chờ» without saying whose turn it is leaves
    both of them waiting for the other.
    """
    who = str(person_id)
    return frozenset(
        str(row["purpose"])
        for row in consents
        if str(row["person_id"]) == who
        and str(row["purpose"]) in CONSENT_PURPOSES
        and _live(row, now=now)
    )


def granted_purposes(
    consents: tuple[dict, ...] | list[dict],
    participants: tuple[str, ...] | list[str],
    *,
    now: datetime | None = None,
) -> frozenset[str]:
    """Purposes BOTH participants have granted and neither has taken back.

    «Both» is the whole rule and it is why this returns a set rather than a
    per-person map: nothing in a two-person notebook is unlocked by one person
    agreeing with themselves.

    Both ON THE SAME PROPOSAL (ADR-0027: «cả hai chấp nhận cùng đề nghị»). Two
    people who each filed their own proposal for one purpose have each agreed
    with themselves; counted per purpose, that read as «both agreed» while no
    proposal had been completed, and the same count gates whether Nếp may read
    the chat (QA 23/09). A row without `proposal_id` belongs to one shared
    group, which is the old per-purpose reading for a caller that has no ids.
    """
    people = {str(p) for p in participants}
    if len(people) < 2:
        return frozenset()
    agreed: dict[tuple[str, object], set[str]] = {}
    for row in consents:
        purpose = str(row["purpose"])
        if purpose not in CONSENT_PURPOSES or not _live(row, now=now):
            continue
        who = str(row["person_id"])
        if who not in people:
            continue
        proposal = row.get("proposal_id")
        key = (purpose, None if proposal is None else str(proposal))
        agreed.setdefault(key, set()).add(who)
    return frozenset(purpose for (purpose, _), who in agreed.items() if who >= people)


def can_bat_doi(
    consents: tuple[dict, ...] | list[dict],
    participants: tuple[str, ...] | list[str],
    *,
    now: datetime | None = None,
) -> bool:
    """Both people have said yes to «Một đôi» (tier 3).

    Tier 2 does not imply this and never will: the database says the same
    thing a second way, with a deferred trigger over `active_couple_members`,
    because this is the rung that makes a person's notebook exclusive.
    """
    return "bat_doi" in granted_purposes(consents, participants, now=now)


def chat_consent_active(
    consents: tuple[dict, ...] | list[dict],
    participants: tuple[str, ...] | list[str],
    *,
    now: datetime,
) -> bool:
    """May Nếp read this pair's messages (tier 4)?

    Asked at the moment of every read, never cached: ADR-0027 says consent is
    checked at every boundary, and a revocation that only takes effect on the
    next restart is not a revocation. The companion asks this before it opens
    the conversation, not after.
    """
    return "doc_chat" in granted_purposes(consents, participants, now=now)


def gu_hai_nguoi(
    consents: tuple[dict, ...] | list[dict],
    participants: tuple[str, ...] | list[str],
    toi: str,
    gu_theo_nguoi: dict[str, list[str]],
    *,
    now: datetime,
) -> dict | None:
    """What of the two tastes one person may see (ADR-0034 §2.1–2.2).

    Only in a notebook both have made «Một đôi»; `None` otherwise, because a
    taste is not something two friends' notebook shares. Within it: the other
    person's tags only if THEY turned `chia_gu` on, and the tags the two have
    in common only if BOTH did -- «common» names the other person's taste too,
    so it needs their yes as much as «theirs» does. My own switch is reported
    so the screen can offer it; my own tags are on my profile already.

    Tags come back in vocabulary order, and a stored tag the vocabulary no
    longer has is left out rather than shown as a raw id.
    """
    people = [str(p) for p in participants]
    me = str(toi)
    if not can_bat_doi(consents, people, now=now):
        return None
    other = next((p for p in people if p != me), None)
    mine_shared = "chia_gu" in granted_by(consents, me, now=now)
    theirs_shared = other is not None and "chia_gu" in granted_by(consents, other, now=now)
    their_tags = set(gu_theo_nguoi.get(other, [])) if theirs_shared and other is not None else set()
    my_tags = set(gu_theo_nguoi.get(me, []))
    theirs = [tag for tag in interests.INTEREST_IDS if tag in their_tags]
    common = [tag for tag in theirs if tag in my_tags] if mine_shared else []
    return {
        "mine_shared": mine_shared,
        "theirs_shared": theirs_shared,
        "theirs": theirs,
        "common": common,
    }


def nguoi_lo_suy(
    participants: list[str],
    to_giay: list[dict],
    *,
    cycle_id: str,
    nguoi_lap_so: str | None,
) -> dict:
    """Who tends to take the lead in this notebook (ADR-0034 §2.3–2.4).

    Read only from what the two did in THIS cycle's notebook, which both
    already hold (ADR-0027 §4): a sheet a person sent first (version 1, by a
    human) counts two, a «đề nghị sửa» they answered with counts one. Nothing
    about who they are -- no gender, no profile, no chat -- goes in.

    The highest score leads. A tie, or nothing yet, goes to whoever opened
    the notebook (`nguoi_lap_so`), and failing that to the first participant.
    `diem` is every participant's score, in participant order, so a screen can
    say why without the rule being restated there.
    """
    people = list(dict.fromkeys(str(p) for p in participants))
    diem = {p: 0 for p in people}
    for to in to_giay:
        if str(to.get("cycle_id")) != str(cycle_id):
            continue
        dau = next((v for v in to.get("versions", ()) if v.get("version") == 1), None)
        if dau is not None and dau.get("author_type") == "human" and dau.get("sent_by") is not None:
            ai = str(dau["sent_by"])
            if ai in diem:
                diem[ai] += 2
        for tl in to.get("responses", ()):
            ai = str(tl.get("person_id"))
            if tl.get("kind") == "de_nghi_sua" and ai in diem:
                diem[ai] += 1
    if not people:
        return {"nguoi_lo": [], "diem": []}
    cao = max(diem.values())
    dau_bang = [p for p in people if diem[p] == cao]
    if len(dau_bang) == 1:
        lo = dau_bang[0]
    elif nguoi_lap_so is not None and str(nguoi_lap_so) in dau_bang:
        lo = str(nguoi_lap_so)
    else:
        lo = dau_bang[0]
    return {"nguoi_lo": [lo], "diem": [[p, diem[p]] for p in people]}


def vai_tuan(suy: dict, chon: dict | None, participants: list[str]) -> dict:
    """This week's «Người lo»: what the two chose for it, or the inference.

    `chon` is the week's stored choice, `{"nguoi_lo_id": id | None}`; None
    there means «Hôm nay mình share», both lead. Choosing is not a permission:
    it decides whose turn the week reads as, nothing else (ADR-0034 §2.4).
    """
    people = list(dict.fromkeys(str(p) for p in participants))
    if chon is None:
        return {"nguoi_lo": list(suy["nguoi_lo"]), "cach": "suy", "diem": suy["diem"]}
    ai = chon.get("nguoi_lo_id")
    return {
        "nguoi_lo": people if ai is None else [str(ai)],
        "cach": "chon",
        "diem": suy["diem"],
    }


def _revision(rows: list[str]) -> str:
    """A short digest of exactly what the preview counted.

    The point is not secrecy, it is that «close» must refuse a preview the
    person read before somebody else changed something: the count they agreed
    to and the rows being closed have to be the same rows. A digest of the
    (id, state) pairs is the cheapest honest way to say «still these».
    """
    material = "\n".join(sorted(rows)).encode("utf-8")
    return hashlib.sha256(material).hexdigest()[:16]


def xem_truoc_dong_so(
    papers: tuple[dict, ...] | list[dict],
    proposals: tuple[dict, ...] | list[dict],
    *,
    now: datetime,
) -> dict:
    """What closing the notebook would do, counted, with a revision to pin it.

    Three numbers, not two. A draft only its owner has seen is DROPPED, a sheet
    waiting for an answer is CANCELLED, and a plan that stands is LOCKED and
    stays readable. The first cut of the client folded the first two together
    and said «1 tờ sẽ khoá» about a draft it then showed as «Đã bỏ» on the very
    next screen; a person reading a destructive confirmation caught it at once
    (blind read, Phase 2). The wire carries all three for that reason.
    """
    nhap_bo = 0
    to_huy = 0
    to_khoa = 0
    material: list[str] = []
    for paper in papers:
        state = pair_paper.hieu_luc(paper, now=now)
        material.append(f"{paper['id']}:{state}")
        if state == "nhap":
            nhap_bo += 1
        elif state in pair_paper.OPEN_STATES:
            to_huy += 1
        elif state in pair_paper.PLAN_STATES:
            to_khoa += 1
    cho = [p for p in proposals if dang_cho(p, now=now)]
    material.extend(f"dn:{p['id']}" for p in cho)
    return {
        "revision": _revision(material),
        "so_nhap_bo": nhap_bo,
        "so_to_huy": to_huy,
        "so_to_khoa": to_khoa,
        "so_de_nghi_huy": len(cho),
    }


def dang_cho(proposal: dict, *, now: datetime) -> bool:
    """An offer still waiting: nobody has completed it and it has not lapsed.

    Public because two readers ask it -- the preview counts offers that closing
    would cancel, and the notebook screen lists the ones a person can still
    answer. Two spellings of «still waiting» is how one screen shows a button
    for something the other has already given up on.
    """
    if proposal.get("completed_at"):
        return False
    expires_at = proposal.get("expires_at")
    if expires_at is not None and now >= expires_at:
        return False
    return True

File: services/core/internal/domain/pairnotebook/pairnotebook.go
// Package pairnotebook is services/api/app/domain/pair_notebook.py: the
// two-person notebook's consent ladder, its cycle vocabulary, and what closing
// it costs.
//
// Pure functions over values. The clock is always a parameter. Python's dicts
// become small structs holding exactly the keys the Python functions read, with
// a nil pointer for a key whose value is None (the service never passes a key
// that is absent, and never a value of another type).
//
// One dependency is not pure in the allowlist's sense: the close preview's
// revision is a SHA-256 digest (hashlib in Python), and crypto/sha256 is not a
// domain import. XemTruocDongSo therefore takes the digest function as a
// parameter; callers pass sha256.Sum256.
//
// The preview reads pair_paper.hieu_luc, OPEN_STATES and PLAN_STATES, as the
// Python module imports them: from the pairpaper package.
package pairnotebook

import (
	"fmt"
	"sort"
	"strings"
	"time"

	"mobile/services/core/internal/domain/interests"
	"mobile/services/core/internal/domain/pairpaper"
)

// OfferWindow is HAN_DE_NGHI: how long an unanswered offer stands.
const OfferWindow = 7 * 24 * time.Hour

// CycleStates is CYCLE_STATES. Each call returns a fresh slice.
func CycleStates() []string { return []string{"pending", "active", "closed"} }

// ConsentPurposes is CONSENT_PURPOSES, the ladder from tier 2 upward, in order.
// `chia_gu` (ADR-0034) is each person's own switch, not a rung both climb.
func ConsentPurposes() []string { return []string{"lap_so", "bat_doi", "doc_chat", "chia_gu"} }

// PerPersonPurposes is PER_PERSON_PURPOSES: what one person decides alone.
func PerPersonPurposes() []string { return []string{"chia_gu"} }

// ConstraintKinds is CONSTRAINT_KINDS.
func ConstraintKinds() []string { return []string{"khong_an_duoc", "dung"} }

// ladder is CONSENT_PURPOSES as the functions below read it.
var ladder = [...]string{"lap_so", "bat_doi", "doc_chat", "chia_gu"}

// NotebookError is NotebookError: a refusal carrying the wire code.
type NotebookError struct {
	Code string
}

func (e *NotebookError) Error() string { return e.Code }

// HanDeNghi is han_de_nghi: when an offer made at now lapses.
//
// Python adds a timedelta to an aware datetime in wall-clock terms of its own
// tzinfo; AddDate does the same in now's Location, which for UTC and fixed
// offsets (all the service ever passes) is also the absolute instant plus a
// week.
func HanDeNghi(now time.Time) time.Time { return now.AddDate(0, 0, 7) }

// Consent is one row of `_consents_as_dicts`: the keys `_live`, `granted_by`
// and `granted_purposes` read. ProposalID "" is Python's absent key: every such
// row belongs to one shared group, the old per-purpose reading.
type Consent struct {
	PersonID            string
	Purpose             string
	GrantedAt           *time.Time
	RevokedAt           *time.Time
	ProposalExpiresAt   *time.Time
	ProposalID          string
	ProposalCompletedAt *time.Time
}

// live is _live: granted, not revoked, and the PROPOSAL not lapsed -- unless it
// was completed: an agreed rung stands until revoked, it does not switch itself
// off when its offer window ends (QA 23/09). A nil now is Python's `now=None`,
// which skips the expiry test entirely.
func live(consent Consent, now *time.Time) bool {
	if consent.GrantedAt == nil {
		return false
	}
	if consent.RevokedAt != nil {
		return false
	}
	if consent.ProposalCompletedAt != nil {
		return true
	}
	if consent.ProposalExpiresAt != nil && now != nil && !now.Before(*consent.ProposalExpiresAt) {
		return false
	}
	return true
}

func onLadder(purpose string) bool {
	for _, rung := range ladder {
		if rung == purpose {
			return true
		}
	}
	return false
}

// GrantedBy is granted_by: what one person has granted and not taken back.
//
// Python returns a frozenset; Go returns its members in ladder order, never
// nil, so equal sets are equal slices.
func GrantedBy(consents []Consent, personID string, now *time.Time) []string {
	held := map[string]bool{}
	for _, row := range consents {
		if row.PersonID == personID && onLadder(row.Purpose) && live(row, now) {
			held[row.Purpose] = true
		}
	}
	out := []string{}
	for _, rung := range ladder {
		if held[rung] {
			out = append(out, rung)
		}
	}
	return out
}

// GrantedPurposes is granted_purposes: the purposes every one of the distinct
// participants has granted ON ONE PROPOSAL (ADR-0027: «cả hai chấp nhận cùng đề
// nghị»). Two people who each filed their own proposal have each agreed only
// with themselves; counted per purpose that read as «both» with nothing
// completed, and the same count gates Nếp reading the chat (QA 23/09). Fewer
// than two distinct people unlock nothing, and duplicates in participants count
// once, as Python's set does.
func GrantedPurposes(consents []Consent, participants []string, now *time.Time) []string {
	people := map[string]bool{}
	for _, person := range participants {
		people[person] = true
	}
	if len(people) < 2 {
		return []string{}
	}
	type key struct{ purpose, proposal string }
	agreed := map[key]map[string]bool{}
	for _, row := range consents {
		if !onLadder(row.Purpose) || !live(row, now) || !people[row.PersonID] {
			continue
		}
		k := key{row.Purpose, row.ProposalID}
		if agreed[k] == nil {
			agreed[k] = map[string]bool{}
		}
		agreed[k][row.PersonID] = true
	}
	held := map[string]bool{}
	for k, who := range agreed {
		if len(who) == len(people) {
			held[k.purpose] = true
		}
	}
	out := []string{}
	for _, rung := range ladder {
		if held[rung] {
			out = append(out, rung)
		}
	}
	return out
}

func contains(values []string, want string) bool {
	for _, value := range values {
		if value == want {
			return true
		}
	}
	return false
}

// CanBatDoi is can_bat_doi: both people have said yes to «Một đôi».
func CanBatDoi(consents []Consent, participants []string, now *time.Time) bool {
	return contains(GrantedPurposes(consents, participants, now), "bat_doi")
}

// ChatConsentActive is chat_consent_active: may Nếp read this pair's messages,
// at now. Python requires the keyword; the service always passes its clock.
func ChatConsentActive(consents []Consent, participants []string, now time.Time) bool {
	return contains(GrantedPurposes(consents, participants, &now), "doc_chat")
}

// Taste is gu_hai_nguoi's dict (ADR-0034 §2.1–2.2).
type Taste struct {
	MineShared   bool
	TheirsShared bool
	Theirs       []string
	Common       []string
}

// GuHaiNguoi is gu_hai_nguoi: what of the two tastes `me` may see. Nil outside
// «Một đôi»; the other's tags only if they turned `chia_gu` on; the common tags
// only if both did. Tags in vocabulary order, unknown ones left out.
func GuHaiNguoi(consents []Consent, participants []string, me string, guTheoNguoi map[string][]string, now time.Time) *Taste {
	if !CanBatDoi(consents, participants, &now) {
		return nil
	}
	var other *string
	for _, person := range participants {
		if person != me {
			other = &person
			break
		}
	}
	mineShared := contains(GrantedBy(consents, me, &now), "chia_gu")
	theirsShared := other != nil && contains(GrantedBy(consents, *other, &now), "chia_gu")
	theirTags := map[string]bool{}
	if theirsShared {
		for _, tag := range guTheoNguoi[*other] {
			theirTags[tag] = true
		}
	}
	myTags := map[string]bool{}
	for _, tag := range guTheoNguoi[me] {
		myTags[tag] = true
	}
	taste := &Taste{MineShared: mineShared, TheirsShared: theirsShared, Theirs: []string{}, Common: []string{}}
	for _, tag := range interests.InterestIDs() {
		if !theirTags[tag] {
			continue
		}
		taste.Theirs = append(taste.Theirs, tag)
		if mineShared && myTags[tag] {
			taste.Common = append(taste.Common, tag)
		}
	}
	return taste
}

// Proposal is one row of `_proposals_as_dicts`.
type Proposal struct {
	ID          string
	CompletedAt *time.Time
	ExpiresAt   *time.Time
}

// DangCho is dang_cho: nobody completed the offer and it has not lapsed.
func DangCho(proposal Proposal, now time.Time) bool {
	if proposal.CompletedAt != nil {
		return false
	}
	if proposal.ExpiresAt != nil && !now.Before(*proposal.ExpiresAt) {
		return false
	}
	return true
}

// Paper is `_paper_dict`, as pair_paper reads it; the preview reads its id,
// state and deadline.
type Paper = pairpaper.Paper

// ClosePreview is xem_truoc_dong_so's dict, keys in the same order.
type ClosePreview struct {
	Revision    string
	SoNhapBo    int
	SoToHuy     int
	SoToKhoa    int
	SoDeNghiHuy int
}

// revision is _revision: the first 16 hex digits of the SHA-256 of the sorted
// material joined by newlines. Python sorts str by code point; for the valid
// UTF-8 every Python str arrives as, byte order is the same order.
func revision(material []string, sum256 func([]byte) [32]byte) string {
	rows := append([]string(nil), material...)
	sort.Strings(rows)
	digest := sum256([]byte(strings.Join(rows, "\n")))
	return fmt.Sprintf("%x", digest[:8])
}

// XemTruocDongSo is xem_truoc_dong_so: what closing would do, counted, with a
// revision that pins exactly the rows counted. sum256 is sha256.Sum256.
func XemTruocDongSo(papers []Paper, proposals []Proposal, now time.Time, sum256 func([]byte) [32]byte) ClosePreview {
	var out ClosePreview
	material := make([]string, 0, len(papers)+len(proposals))
	for _, paper := range papers {
		state := pairpaper.HieuLuc(paper, now)
		material = append(material, paper.ID+":"+state)
		switch {
		case state == "nhap":
			out.SoNhapBo++
		case pairpaper.IsOpen(state):
			out.SoToHuy++
		case pairpaper.IsPlan(state):
			out.SoToKhoa++
		}
	}
	for _, proposal := range proposals {
		if DangCho(proposal, now) {
			material = append(material, "dn:"+proposal.ID)
			out.SoDeNghiHuy++
		}
	}
	out.Revision = revision(material, sum256)
	return out
}

// ToTinHieu is `_paper_signals`: what nguoi_lo_suy reads of one sheet. A nil
// CycleID is Python's None; a Version with a nil SentBy was never sent.
type ToTinHieu struct {
	CycleID   *string
	Versions  []PhienBanTinHieu
	Responses []TraLoiTinHieu
}

// PhienBanTinHieu is one version as nguoi_lo_suy reads it.
type PhienBanTinHieu struct {
	Version    int
	AuthorType string
	SentBy     *string
}

// TraLoiTinHieu is one response as nguoi_lo_suy reads it.
type TraLoiTinHieu struct {
	PersonID string
	Kind     string
}

// Diem is one participant's score.
type Diem struct {
	PersonID string
	Score    int
}

// NguoiLo is nguoi_lo_suy's and vai_tuan's dict: NguoiLo lists who leads,
// Cach is "suy" or "chon" ("" from nguoi_lo_suy alone).
type NguoiLo struct {
	NguoiLo []string
	Cach    string
	Diem    []Diem
}

// NguoiLoSuy is nguoi_lo_suy (ADR-0034 §2.3–2.4): who tends to take the lead,
// read only from this cycle's sheets; a tie or nothing yet goes to whoever
// opened the notebook, then to the first participant.
func NguoiLoSuy(participants []string, toGiay []ToTinHieu, cycleID string, nguoiLapSo *string) NguoiLo {
	people := []string{}
	seen := map[string]bool{}
	for _, p := range participants {
		if !seen[p] {
			seen[p] = true
			people = append(people, p)
		}
	}
	diem := map[string]int{}
	for _, to := range toGiay {
		if to.CycleID == nil || *to.CycleID != cycleID {
			continue
		}
		for _, v := range to.Versions {
			if v.Version != 1 {
				continue
			}
			if v.AuthorType == "human" && v.SentBy != nil && seen[*v.SentBy] {
				diem[*v.SentBy] += 2
			}
			break
		}
		for _, tl := range to.Responses {
			if tl.Kind == "de_nghi_sua" && seen[tl.PersonID] {
				diem[tl.PersonID]++
			}
		}
	}
	if len(people) == 0 {
		return NguoiLo{NguoiLo: []string{}, Diem: []Diem{}}
	}
	cao := diem[people[0]]
	for _, p := range people {
		if diem[p] > cao {
			cao = diem[p]
		}
	}
	dauBang := []string{}
	for _, p := range people {
		if diem[p] == cao {
			dauBang = append(dauBang, p)
		}
	}
	lo := dauBang[0]
	if len(dauBang) > 1 && nguoiLapSo != nil && contains(dauBang, *nguoiLapSo) {
		lo = *nguoiLapSo
	}
	out := NguoiLo{NguoiLo: []string{lo}, Diem: []Diem{}}
	for _, p := range people {
		out.Diem = append(out.Diem, Diem{PersonID: p, Score: diem[p]})
	}
	return out
}

// VaiTuan is vai_tuan: the week's stored choice, or the inference. chon nil
// is Python's None (nothing chosen); chon pointing at nil is «cả hai».
func VaiTuan(suy NguoiLo, chon **string, participants []string) NguoiLo {
	if chon == nil {
		return NguoiLo{NguoiLo: append([]string{}, suy.NguoiLo...), Cach: "suy", Diem: suy.Diem}
	}
	if *chon == nil {
		people := []string{}
		seen := map[string]bool{}
		for _, p := range participants {
			if !seen[p] {
				seen[p] = true
				people = append(people, p)
			}
		}
		return NguoiLo{NguoiLo: people, Cach: "chon", Diem: suy.Diem}
	}
	return NguoiLo{NguoiLo: []string{**chon}, Cach: "chon", Diem: suy.Diem}
}
